apps/browser/bookmarks/bookmarkfiltermodel.cpp

```diff
@@ -28,24 +28,53 @@ bool BookmarkFilterModel::filterAcceptsRow(int sourceRow, const QModelIndex &sou
 {
     QModelIndex index = sourceModel()->index(sourceRow, 0, sourceParent);
 
-    if (sourceModel()->data(index, DeclarativeBookmarkModel::UrlRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)
-            || sourceModel()->data(index, DeclarativeBookmarkModel::TitleRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)) {
+    if ((sourceModel()->data(index, DeclarativeBookmarkModel::UrlRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)
+         || sourceModel()->data(index, DeclarativeBookmarkModel::TitleRole).toString().trimmed().contains(search(), Qt::CaseInsensitive))
+         && m_countFilterAccepts < m_maxDisplayedItems) {
+        m_countFilterAccepts++;
         return true;
     }
     return false;
 }
 
+void BookmarkFilterModel::setSourceModel(QAbstractItemModel *sourceModel)
+{
+    if (sourceModel) {
+        m_countFilterAccepts = 0;
+        m_maxDisplayedItems = sourceModel->rowCount();
+        QSortFilterProxyModel::setSourceModel(sourceModel);
+        invalidate();
+    }
+}
+
 QString BookmarkFilterModel::search() const
 {
     return m_search;
 }
 
+int BookmarkFilterModel::maxDisplayedItems() const
+{
+    return m_maxDisplayedItems;
+}
+
 void BookmarkFilterModel::setSearch(const QString &search)
 {
     if (m_search == search)
         return;
 
+    m_countFilterAccepts = 0;
     m_search = search;
+    invalidate();
     emit searchChanged(m_search);
+}
+
+void BookmarkFilterModel::setMaxDisplayedItems(const int maxDisplayedItems)
+{
+    if (m_maxDisplayedItems == maxDisplayedItems)
+        return;
+
+    m_countFilterAccepts = 0;
+    m_maxDisplayedItems = maxDisplayedItems;
     invalidate();
```

Comment:
Similarly, if `maxDisplayedItems` changes you will only ever be adding/remove rows to/from the end of the model, so it would be better not to invalidate the entire model.

Reply:
It looks like `invalidateFilter()` would apply here too.

Reply:
Done